### ISSUE TITLE
Implement standard and configurable gestures across all platforms

### ISSUE THREAD

**Issue by madmiraal** [godotengine/godot-proposals#4340]
### Describe the project you are working on

Creating interactive UIs including the Godot Editor.

### Describe the problem or limitation you are having in your project

Godot currently has two gesture events: `InputEventPanGesture` and `InputEventMagnifyGesture`. `InputEventPanGesture` is generated by macOS and Android's “scroll” gestures. `InputEventMagnifyGesture` is generated by macOS' “pinch” gesture. Within the editor, `InputEventPanGesture` is used by: `ItemList`, `PopupMenu` (in 3.x) and `RichTextLabel` for scrolling and by `ScrollContainer`, `GraphEdit`, `TextEdit` and `Tree` for panning. Internally, the code editor uses `InputEventMagnifyGesture` to change the font size. The canvas item editor uses both to control zooming. The polygon 2D editor uses both to control zooming and panning. And the node 3D editor uses both to zoom, pan and rotate. However, in general, Godot currently uses hard-coded combinations of actions instead of gestures to produce outcomes, which leads to inconsistencies (see #1507).

`InputEventPanGesture` and `InputEventMagnifyGesture` were added in 2017 in godotengine/godot#12573 as _"A bit of an experiment"_ on macOS. Since godotengine/godot#12573 was merged, godotengine/godot#13139 has been open to implement these gestures on other platforms. godotengine/godot#25474 attempted to add these gestures to Android, but it was reverted in godotengine/godot#33536. godotengine/godot#36953 was opened to try again to implement gestures in Android. However there is also a platform independent approach: godotengine/godot#39055 and its 3.x version godotengine/godot#37754.

The problem is there is a difference between inputs, gestures and outcomes, and these concepts are often used interchangeably, which causes confusion. Inputs are the events generated by the user interacting with devices e.g. a mouse button being pressed or a screen being touched. Gestures are combinations of events that occur over time and are generally directed at something e.g. clicking the mouse over an item or tapping an item on the screen. Finally, users expect their gestures to have an expected outcome e.g. the item is selected.

While inputs are definitive, gestures are not. There are some de facto gestures e.g. clicking the mouse button. However, even with this gesture it is not agreed whether the click happens on the mouse button down or up (see Godot’s button’s [`ActionMode`](https://docs.godotengine.org/en/stable/classes/class_basebutton.html#enum-basebutton-actionmode)). Furthermore, the temporal nature of gestures and the differences between users’ capabilities makes defining them impossible e.g. the time and movement between two clicks differentiates between two single clicks and a double click. To accommodate different users, most operating systems make the maximum time and movement between clicks to define a double-click configurable. And these are just the problems with clicking and double-clicking, never mind more complex gestures. Therefore, any gestures we define in Godot should have configurable parameters.

Combinations of events from different sources are also expected to generate the same gestures, for example, clicking the left-mouse button or tapping the screen. The [emulate mouse from touch](https://docs.godotengine.org/en/stable/classes/class_projectsettings.html#class-projectsettings-property-input-devices-pointing-emulate-mouse-from-touch) and [emulate touch from mouse](https://docs.godotengine.org/en/stable/classes/class_projectsettings.html#class-projectsettings-property-input-devices-pointing-emulate-touch-from-mouse) project settings are used to try and workaround this issue, but this leaves issues like godotengine/godot#24589.

What outcome a gesture should have or what gesture or multiple gestures should produce a desired outcome is even more subjective. Even the press and move gesture has multiple, conflicting interpretations in the engine: box select or select and move (See #776). Therefore, regardless of what gestures are defined in Godot, each game should be responsible for the outcomes and what gestures are used to generate them. Within the engine, users should be able to decide what, if any, gesture produces an outcome.

The question is: What gestures should be defined in Godot?

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Godot should implement a minimum set of standard gestures in the `Input` system and remove existing hard-coded gestures used by the platforms and defined in the engine code.

The solution should allow users to:
- Configure the parameters of these gestures.
- Add these gestures to `Action`s in the `InputMap` system.
- Create their own gestures.

All platforms send their input events to the input system. The input system would combine these events to create the standard, expected gestures. Different combinations of input events from different devices can create the same gesture. Gestures should be named to reflect the action not the expected outcome. It’s worth point out that “Scroll”, “Pan”, “Magnify” and “Twist” are all outcomes not actions. To enable users to configure these gestures, the gesture parameters would be available in the `Project Settings`.

The following are a suggested minimum set of expected gestures:
- Click (or tap)
- Double-click (or tap)
- Triple-click (or tap)
- Long press
- Click and Drag (the current pan gesture)
- Pinch (or some other name that better describes two fingers click and dragging independently)
- Swipe (a touch, drag and lift with time constraints).
- Long-press and drag

For brevity, I won't detail the gestures' (configurable) parameters or their properties. These details should be reasonably self-explanatory, but should be agreed too.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

The best way to show how this will work will be through a PR. In the meantime, although I disagree with the details, and it doesn't include a lot of what is described here, godotengine/godot#39055 provides a good idea of the approach.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Gestures are a combination of events, which users can use to create their own gestures in scripts. Combining events to create hard-coded gestures is the general approach used within the engine and OS code. However, as described above, this leads to inconsistencies and, when it's hard-coded it's inflexible.

### Is there a reason why this should be core and not an add-on in the asset library?

Users have already implemented their own gestures as addons. For example: [Godot Touch Input Manager](https://godotengine.org/asset-library/asset/473). However, there are some standard gestures that users and the engine expect to be defined. However, there are a minimum set of gestures that users expect to be available. Furthermore, there are gestures used within the editor or users expect to be available within the editor that need to be available internally.

**Comment by grayhaze** [comment]
I agree with everything proposed here, except that maybe Pinch could just be Multi-touch, as there's no real need to limit it to two touches. Using the correct math, it should be possible to implement pinching and twisting using any number of touches, and the gesture could provide the number of touches in case different actions were wanted for different numbers.

**Comment by novalis** [comment]
Gestures are a UI language, and the meaning of gestures is often platform-specific.  A mouse click is not a tap, because a long-click is (on Windows and Linux) meaningless, while a long-tap on a mobile device often opens a context menu. 

Even among mobile platforms, gestures are not standard. It's true that Android and iOS both have pinch-to-zoom.  But In addition to the gestures that Android has swiped, iOS also has three-finger-swipe-left for undo (also, shake-to-undo).  That's the thing that brought me to this proposal: I want to add undo to my game and have it be idiomatic.  That's impossible on Android: there is no undo gesture.  But it's possible on Windows and Linux (ctrl-z) and iOS (the gesture).  It's also possible in HTML5, using [horrible hackery](https://dev.to/chromiumdev/-native-undo--redo-for-the-web-3fl3).  

Godot already has some of what @madmiraal would define as outcome-based events (`ui_select`, or `NOTIFICATION_WM_GO_BACK_REQUEST`).  I think of these as "intent"-based rather than "outcome"-based.  That is, the user intends to select the thing (or undo, or go back, or whatever).  That may or may not work: you can't undo before you've done something, for instance.  

So I think we should add a few more of these for the standard gestures (using the operating system's gesture recognizer, where available), so that we can have idiomatic controls, instead of trying to standardize across platforms.

**Comment by pro-rookie-gamer** [comment]
Will this allow me to make the editor's 3D navigation more similar to Blender's?

**Comment by esotericpig** [comment]
Would this also be added for Mobile/Tablet Web Browsers? Currently, I have to do this hack for pinch zoom on mobile web browsers. Others can feel free to copy & use this code, no problem.

```gdscript
# Could be an array instead and modify methods to check bounds.
var _touch_index_to_pos: Dictionary[int,Vector2] = {}
var _last_pinch_dist := 0.0

func _unhandled_input(event: InputEvent) -> void:
  if event is InputEventMagnifyGesture:
    print('This does not get called on mobile web browsers.')
  elif event is InputEventScreenTouch:
    var e := event as InputEventScreenTouch

    _set_pinch_hack(e.is_pressed(),e.index,e.position)
  elif event is InputEventScreenDrag:
    var e := event as InputEventScreenDrag

    if !e.is_canceled():
      _pinch_hack(e.index,e.position)

func _set_pinch_hack(is_pressed: bool,touch_index: int,touch_pos: Vector2) -> void:
  if is_pressed:
    _touch_index_to_pos[touch_index] = touch_pos
  else:
    _touch_index_to_pos.erase(touch_index)

func _pinch_hack(touch_index: int,touch_pos: Vector2) -> void:
  _touch_index_to_pos[touch_index] = touch_pos

  if !_is_pinching():
    return

  var pos1 := _touch_index_to_pos[0]
  var pos2 := _touch_index_to_pos[1]
  var pinch_dist := pos1.distance_to(pos2)

  if _last_pinch_dist > 0.0:
    var factor := pinch_dist / _last_pinch_dist
    var center_gpos := global_position + ((pos1 + pos2) / 2.0)

    # TODO: zoom on center_gpos using factor

  _last_pinch_dist = pinch_dist

# Use this to ignore taps while pinching.
func _is_pinching() -> bool:
  return (_touch_index_to_pos.size() >= 2 &&
          (_touch_index_to_pos.has(0) && _touch_index_to_pos.has(1)))
```

**Comment by DigitalTableTops** [comment]
Thank you [esotericpig](https://github.com/esotericpig). I have been having trouble implementing InputEventMagnifyGesture on mobile browsers. I wasn't sure if it was just me or not. I was going to do some more testing and maybe open a new issue for this. But it seems like this is something that people are aware of.

In the meantime, I will try your code. It looks like it should do the trick. Thanks again!
